src/tests.js

```diff
@@ -0,0 +1,248 @@
+								eventTestId: data.testId,
+								eventPostId: data.postId,
+								eventVariantId: data.variantId,
+							},
+							metrics: {
+								...metrics,
+							}
+						} );
+					} );
+				} );
+			} );
+		}
+
+		getVariantId () {
+			const testId = `${ this.testId }_${ this.postId }`;
+			const trafficPercentage = this.trafficPercentage;
+
+			// Extract test set by campaign parameter and add for user.
+			const utm_test = window.location.search.match( /utm_campaign=test_([a-z0-9_-]+):(\d+)/i );
+			if ( utm_test ) {
+				this.addTestForUser( { [ utm_test[ 1 ] ]: parseInt( utm_test[ 2 ], 10 ) } );
+			}
+
+			// Check if this user already have a variant for this test.
+			const currentTests = this.getTestsForUser();
+			let variantId = false;
+			// Test variant can be 0 so check for not undefined and not strictly false and
+			// that it's a valid index.
+			if (
+				typeof currentTests[ testId ] !== 'undefined' &&
+				currentTests[ testId ] !== false &&
+				currentTests[ testId ] < this.variantCount
+			) {
+				variantId = currentTests[ testId ];
+			} else if ( currentTests[ testId ] === false ) {
+				return variantId;
+			} else {
+				// Otherwise lets check the probability we should experiment on this individual.
+				// That sounded weird.
+				if ( Math.random() * 100 > trafficPercentage ) {
+					// Exclude from this test.
+					this.addTestForUser( {
+						[ testId ]: false
+					} );
+					return variantId;
+				}
+				// Add one of the variants to the cookie.
+				variantId = Math.floor( Math.random() * this.variantCount );
+				this.addTestForUser( {
+					[ testId ]: variantId
+				} );
+			}
+
+			// Log active test variant for all events.
+			if ( window.Altis && window.Altis.Analytics ) {
+				window.Altis.Analytics.registerAttribute( `test_${ testId }`, variantId );
+			}
+
+			return variantId;
+		}
+
+		getTestsForUser () {
+			return JSON.parse( window.localStorage.getItem( "_altis_ab_tests" ) ) || {};
+		}
+
+		addTestForUser ( test ) {
+			window.localStorage.setItem( "_altis_ab_tests", JSON.stringify( {
+				...this.getTestsForUser(),
+				...test
+			} ) );
+		}
+
+	}
+
+	ABTest.goalHandlers = {};
+
+	/**
+	 * Add an event handler for recording an analytics event.
+	 * The event is then used to determine the goal success.
+	 *
+	 * Callback receives the target node and a function to record
+	 * to record the event.
+	 *
+	 * @param
+	 */
+	ABTest.registerGoal = ( name, callback, closest = [] ) => {
+		ABTest.goalHandlers[ name ] = {
+			callback,
+			closest: Array.isArray( closest ) ? closest : [ closest ],
+		};
+	};
+
+	// Register built in click goal handler.
+	ABTest.registerGoal( 'click', ( element, record ) => {
+		// Collect attributes.
+		const attributes = {
+			sourceNode: element.nodeName || '',
+			sourceText: element.innerText || '',
+			sourceClassName: element.className || '',
+			sourceId: element.id || '',
+		};
+
+		// Bind handler.
+		element.addEventListener( 'click', event => {
```

Comment:
I'm wondering if we need to check if the element is inside an `<a>` tag, or maybe _is_ / has an `<a>` tag. For the sake of titles, if you just output the title, and it gets AB-swapped, but there's no link, would be want clicking that text to trigger this?

Reply:
@joehoyle yep I considered that - it's a little odd but the best way I could think to do it. When the goal is registered we pass a 3rd parameter `closest` which is an array of allowed tag names. By default it will work up the DOM tree to find the nearest allowed node and pass that as the element to the callback :)